sklearn/preprocessing/tests/test_target_encoder.py

```diff
@@ -0,0 +1,308 @@
+            [[0, 1, 2]],
+        ),
+        (
+            np.array(
+                [["cat"] * 10 + ["dog"] * 10 + ["snake"]], dtype=object
+            ).T,  # snake is unknown
+            [["dog", "cat", "cow"]],
+        ),
+    ],
+)
+@pytest.mark.parametrize("smooth", [4.0, "auto"])
+def test_custom_categories(X, categories, smooth):
+    """Custom categories with unknown categories that are not in training data."""
+    rng = np.random.RandomState(42)
+    y = rng.uniform(low=-10, high=20, size=X.shape[0])
+    enc = TargetEncoder(categories=categories, smooth=smooth).fit(X, y)
+
+    # The last element is unknown and encoded as the mean
+    y_mean = y.mean()
+    X_trans = enc.transform(X[-1:])
+    assert X_trans[0, 0] == pytest.approx(y_mean)
+
+    assert len(enc.encodings_) == 1
+    # custom category that is not in training data
+    assert enc.encodings_[0][-1] == pytest.approx(y_mean)
+
+
+@pytest.mark.parametrize(
+    "y, msg",
+    [
+        ([1, 2, 0, 1], "Found input variables with inconsistent"),
+        (
+            np.array([[1, 2, 0], [1, 2, 3]]).T,
+            "Target type was inferred to be 'multiclass-multioutput'",
+        ),
+        (["cat", "dog", "bear"], "Target type was inferred to be 'multiclass'"),
+    ],
+)
+def test_errors(y, msg):
+    """Check invalidate input."""
+    X = np.array([[1, 0, 1]]).T
+
+    enc = TargetEncoder()
+    with pytest.raises(ValueError, match=msg):
+        enc.fit_transform(X, y)
+
+
+def test_use_regression_target():
+    """Custom target_type to avoid inferring the target type."""
+    X = np.array([[0, 1, 0, 1, 0, 1]]).T
+
+    # XXX: When multiclass is supported, then the following `y`
+    # is considered a multiclass problem and `TargetEncoder` will not error.
+    # type_of_target would be 'multiclass'
+    y = np.array([1.0, 2.0, 3.0, 2.0, 3.0, 4.0])
+    enc = TargetEncoder()
+    msg = "Target type was inferred to be 'multiclass'"
+    with pytest.raises(ValueError, match=msg):
+        enc.fit_transform(X, y)
+
+    enc = TargetEncoder(target_type="continuous")
+    enc.fit_transform(X, y)
+    assert enc.target_type_ == "continuous"
+
+
+def test_feature_names_out_set_output():
+    """Check TargetEncoder works with set_output."""
+    pd = pytest.importorskip("pandas")
+
+    X_df = pd.DataFrame({"A": ["a", "b"] * 10, "B": [1, 2] * 10})
+    y = [1, 2] * 10
+
+    enc_default = TargetEncoder(cv=2, smooth=3.0).set_output(transform="default")
+    enc_pandas = TargetEncoder(cv=2, smooth=3.0).set_output(transform="pandas")
+
+    X_default = enc_default.fit_transform(X_df, y)
+    X_pandas = enc_pandas.fit_transform(X_df, y)
+
+    assert_allclose(X_pandas.to_numpy(), X_default)
+    assert_array_equal(enc_pandas.get_feature_names_out(), ["A", "B"])
+    assert_array_equal(enc_pandas.get_feature_names_out(), X_pandas.columns)
+
+
+@pytest.mark.parametrize("to_pandas", [True, False])
```

Comment:
I am certainly biased but I liked to use `_convert_container` from the `testing` utils.

Reply:
For this specific test, if `to_pandas=True`, then the second feature is converted into a strings improve test coverage. I do not see a way to add `_convert_container` without complicating the logic in:

https://github.com/thomasjpfan/scikit-learn/blob/acb4b2eb4fea92ef0ee6507eb3c5de713a9bafd0/sklearn/preprocessing/tests/test_target_encoder.py#L235-L244

I guess it can become the following:

```python
    if constructor_name == "dataframe":
        # convert second feature to an object
        X_input = _convert_container(
            {
                "feat0": X_int[:, 0],
                "feat1": np.array(["cat", "dog"], dtype=object)[X_int[:, 1]],
            },
            constructor_name,
        )
        # "snake" is unknown
        X_test = _convert_container(
            {"feat0": X_test[:, 0], "feat1": ["dog", "cat", "snake"]}, constructor_name
        )
```

But I feel it adds a little bit too much indirection.

Reply:
OK. Fine then.